File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.cc
#include <iostream>
#include <math.h>
using namespace std;
int n, resulta = 0, resultb = 0, total = 0, balance = 2000;
int stat[20][20] = {0, };
bool team[10] = {0,};
void dfs(int cnt, int num){
    if(cnt == n/2){
        for(int i = 0; i < n; i++){
            if(team[i]){
                for(int j = 0; j < n; j ++){
                    if(team[j]){
                        resulta += stat[i][j];
                    }
                }
            }
        }
        for(int i = 0; i < n; i++){
            if(!team[i]){
                for(int j = 0; j < n; j ++){
                    if(!team[j]){
                        resultb += stat[i][j];
                    }
                }
            }
        }

        if(abs(resulta -resultb)< balance){ 
            balance = abs(resulta - resultb);

        }
        resulta = 0;
        resultb  = 0;
    }
    for(int i = num; i < n; i++){
        if(!team[i]){
            team[i] = true;
            dfs(cnt+1, i+1);
            team[i] = false;
        }

    }
    
    
    
}

int main(){
    int m;
    cin >> n;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n ; j++){
            cin >> m;
            stat[i][j] = m;
            total += m;
        }
    }
    dfs(0,0);
    cout << balance;
    
}

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Main {
	static class Team implements Comparable<Team> {
		BigInteger players;
		int stat;

		public Team(BigInteger players, int stat) {
			this.players = players;
			this.stat = stat;
		}

		@Override
		public int hashCode() {
			return this.players.hashCode();
		}

		@Override
		public int compareTo(Team o) {
			return this.players.compareTo(o.players);
		}
	}

	public static void main(String[] args) throws Exception {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(in.readLine());
		int[][] table = new int[n][n];
		String maxStr = "";
		for (int i = 0; i < n; i++) {
			String[] split = in.readLine().split(" ");
			maxStr += "1";
			for (int j = 0; j < n; j++) {
				table[i][j] = Integer.parseInt(split[j]);
			}
		}
		int half = n / 2;
		Set<BigInteger>[] dp = new HashSet[half + 1];
		for (int i = 0; i <= half; i++) {
			dp[i] = new HashSet<>();
		}
		BigInteger zero = new BigInteger("0");
		dp[0].add(new BigInteger("0"));
		for (int i = 0; i < half; i++) {
			Set<BigInteger> tList = dp[i];
			for (int j = 0; j < n; j++) {
				for (BigInteger t : tList) {
					if (!t.testBit(j)) {
						BigInteger nextP = t.setBit(j);
						dp[i + 1].add(nextP);
					}

				}
			}
		}

		int ans = Integer.MAX_VALUE;
		BigInteger max = new BigInteger(maxStr, 2);
		Set<BigInteger> completeTeam = dp[half];
		Set<BigInteger> visit = new HashSet<>();
		for (BigInteger t1 : completeTeam) {
			BigInteger t2 = max.xor(t1);
			if(visit.add(t1)&&visit.add(t2)) {
				int t1Stat = 0;
				int t2Stat = 0;
				for(int i = 0;i<t1.bitLength();i++) {
					if(t1.testBit(i)) {
						for(int j = i+1;j<t1.bitLength();j++) {
							if(t1.testBit(j)) {
								t1Stat+= table[i][j]+table[j][i];
							}
						}
					}
				}
				for(int i = 0;i<t2.bitLength();i++) {
					if(t2.testBit(i)) {
						for(int j = i+1;j<t2.bitLength();j++) {
							if(t2.testBit(j)) {
								t2Stat+= table[i][j]+table[j][i];
							}
						}
					}
				}
				int diff = Math.abs(t1Stat - t2Stat);
				if (ans > diff) {
					ans = diff;
				}
			}
		}

		System.out.println(ans);

	}

}